File: Packages/Yage/Sources/Yage/Models/Entity.swift
import Schwifty
import SwiftUI

open class Entity: Identifiable {
    public let species: Species
    public let id: String
    public var capabilities: [Capability] = []
    public var direction: CGVector = .zero
    public var fps: TimeInterval = 10
    public var frame: CGRect
    public var isAlive = true
    public var isEphemeral: Bool = false
    public var isStatic: Bool = false
    public var speed: CGFloat = 0
    public var sprite: String?
    public private(set) var state: EntityState = .move
    public var worldBounds: CGRect
    public var zIndex: Int

    public init(
        species: Species,
        id: String,
        frame: CGRect,
        in worldBounds: CGRect
    ) {
        self.species = species
        self.id = id
        self.fps = species.fps
        self.zIndex = species.zIndex
        self.frame = frame
        self.worldBounds = worldBounds
        self.installCapabilities()
    }
    
    private func installCapabilities() {
        species.capabilities.forEach {
            if let capability = CapabilitiesDiscoveryService.shared.capability(for: $0) {
                install(capability)
            }
        }
    }

    // MARK: - Capabilities

    public func capability<T: Capability>(for someType: T.Type) -> T? {
        capabilities.first { $0 as? T != nil } as? T
    }

    // MARK: - Update

    open func update(with collisions: Collisions, after time: TimeInterval) {
        guard isAlive else { return }
        capabilities.forEach {
            $0.update(with: collisions, after: time)
        }
    }

    // MARK: - State

    open func set(state newState: EntityState) {
        state = newState
        Logger.log(id, "State changed to", state.description)
    }

    // MARK: - Memory Management

    open func kill() {
        uninstallAllCapabilities()
        isAlive = false
        sprite = nil
    }

    public func uninstallAllCapabilities() {
        capabilities.forEach { $0.kill(autoremove: false) }
        capabilities = []
    }
}

// MARK: - Debug

extension Entity: CustomStringConvertible {
    public var description: String {
        id
    }
}

// MARK: - Equatable

extension Entity: Equatable {
    public static func == (lhs: Entity, rhs: Entity) -> Bool {
        lhs.id == rhs.id
    }
}

File: Sources/windows/yage/models/entity.py
from datetime import datetime
from typing import Any, List, Optional, Tuple
from di import Dependencies
from yage.models.animations import EntityAnimation
from yage.models.capability import CapabilitiesDiscoveryService
from yage.models.collisions import Collision
from yage.models.entity_state import EntityState
from yage.utils.geometry import Vector
from yage.utils.logger import Logger


class Entity:
    def __init__(self, species, entity_id: str, frame, world):
        self._animation = None
        self._animation_loops = None
        self.capabilities = []
        self.creation_date = datetime.now()
        self.direction = Vector.zero()
        self.fps = species.fps
        self.frame = frame
        self.id = entity_id
        self.is_alive = True
        self.is_ephemeral = False
        self.is_static = False
        self.species = species
        self.speed = 0
        self.sprite = None
        self.state = EntityState.MOVE
        self.world = world
        self.z_index = 0
        self._install_capabilities()

    @property
    def world_bounds(self):
        return self.world.bounds

    def animation(self) -> Tuple['EntityAnimation', int]:
        if self.state != EntityState.ANIMATION:
            return (None, None)
        return (self._animation, self._animation_loops)

    def set_animation(self, animation: 'EntityAnimation', loops: int):
        self._animation = animation
        self._animation_loops = loops
        self.state = EntityState.ANIMATION

    def collisions(self, neighbors: List['Entity']) -> List['Collision']:
        valid_neighbors = [n for n in neighbors if n != self]
        collisions = [self.collision(n) for n in valid_neighbors]
        collisions = [c for c in collisions if c is not None]
        return collisions

    def collision(self, other: 'Entity') -> Optional['Collision']:
        intersection = self.frame.intersection(other.frame)
        if not intersection:
            return None
        return Collision(self, other, intersection)

    def __setattr__(self, name: str, value: Any) -> None:
        super().__setattr__(name, value)
        if name == 'state':
            Logger.log(self.id, 'State changed to', self.state)
            if self.state == EntityState.MOVE:
                self.reset_speed()

    def _install_capabilities(self):
        for capability_id in self.species.capabilities:
            capability = Dependencies.instance(
                CapabilitiesDiscoveryService).capability(capability_id)
            if capability is not None:
                self.install(capability)

    def reset_speed(self):
        self.speed = 1

    def install(self, capability_type):
        capability = capability_type(self)
        self.capabilities.append(capability)
        return capability

    def capability(self, some_type):
        for capability in self.capabilities:
            if isinstance(capability, some_type):
                return capability
        return None

    def update(self, collisions, time):
        if not self.is_alive:
            return
        for capability in self.capabilities:
            capability.update(collisions, time)

    def kill(self):
        self._uninstall_all_capabilities()
        self.is_alive = False
        self.sprite = None

    def _uninstall_all_capabilities(self):
        for capability in self.capabilities:
            capability.kill(False)
        self.capabilities = []

    def __repr__(self):
        return self.id

    def __eq__(self, other):
        return self.id == other.id
